[FSDP2] idempotent reset_sharded_param: no-op if _local_tensor is already padded

weifengpy opened the pull request:
resolves https://github.com/pytorch/torchtitan/issues/1136

torchtitan use cached state dict for ft. reset_sharded_param should be idempotent if model.parameters() are padded already

```
# pad DTensor._local_tensor
fully_shard(model)
sd = fsdp_model.state_dict()
# reset_sharded_param should be a no-op in lazy_init
loss = fsdp_model(inp).sum()
```

this PR make `reset_sharded_param` idempotent by checking storage data ptr and return early

unit test
```
pytest -s test/distributed/_composable/fsdp/test_fully_shard_state_dict.py -k test_cached_state_dict
```
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #163130



cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim @dcci

weifengpy committed:
reset sharded params in state_dict() and make it idempotent

[ghstack-poisoned]

weifengpy committed:
Update on "reset sharded params in state_dict() and make it idempotent"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

weifengpy committed:
Update on "[FSDP2] reset sharded params in state_dict() and make it idempotent"

resolves https://github.com/pytorch/torchtitan/issues/1136




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

weifengpy committed:
Update on "[FSDP2] reset sharded params in state_dict() and make it idempotent"

resolves https://github.com/pytorch/torchtitan/issues/1136




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

weifengpy commented:
ci error is not relevant

fegin commented:
@awgu The current issue is that some training frameworks, including TorchTitan, calls `model.state_dict()` before the first `forward()` and uses that result throughout the entire training without calling `model.state_dict()` again. This will result in incorrect checkpoints being saved.

tianyu-l reviewed (approved):
Thanks for the fix!

weifengpy committed:
Update on "[FSDP2] reset sharded params in state_dict() and make it idempotent"

resolves https://github.com/pytorch/torchtitan/issues/1136

torchtitan use cached state dict for ft. fsdp2 should run padding for sharded params
```
# should call reset sharded params for padding
sd = fsdp_model.state_dict()
# reset sharded params should be a no-op
loss = fsdp_model(inp).sum()
```

this PR does two thing
* reset sharded params in state dict pre hook 
* make sharded params idempotent by checking storage data ptr and return early


unit test
```
pytest -s test/distributed/_composable/fsdp/test_fully_shard_state_dict.py -k test_cached_state_dict
```



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

weifengpy commented:
> where did the padding get lost in the first place / why was there not padding re-added at that point

good question! no padding is getting losted. here is what I want to achieve

fully_shard(model) with padded local_tensor  -> model.state_dict() -> model(input) **reset_sharded_param should be no-op**

I just need reset_sharded_param to be idempotent. without the PR, we are always creating new padded tensors, because `local_tensor.size() != padded_sharded_size` is always true

No need to call reset_sharded_param in state dict hooks. I modified the PR

weifengpy commented on torch/distributed/fsdp/_fully_shard/_fsdp_param.py:
```diff
@@ -833,10 +833,20 @@ def reset_sharded_param(self):
             return
         updated_local_tensor = False
+        # `reset_sharded_param` can be called twice
+        # 1st time in sd = model.state_dict()
```
updated. good catch!

weifengpy committed:
Update on "[FSDP2] reset sharded params in state_dict() and make it idempotent"

resolves https://github.com/pytorch/torchtitan/issues/1136

torchtitan use cached state dict for ft. reset_sharded_param should be idempotent if model.parameters() are padded already

```
# pad DTensor._local_tensor
fully_shard(model)
sd = fsdp_model.state_dict()
# reset_sharded_param should be a no-op in lazy_init
loss = fsdp_model(inp).sum()
```

this PR make `reset_sharded_param` idempotent by checking storage data ptr and return early

unit test
```
pytest -s test/distributed/_composable/fsdp/test_fully_shard_state_dict.py -k test_cached_state_dict
```



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

weifengpy commented:
@pytorchmergebot merge

fegin commented:
@weifengpy The test failure looks real.

weifengpy committed:
Update on "[FSDP2] idempotent reset_sharded_param: no-op if _local_tensor is already padded"

resolves https://github.com/pytorch/torchtitan/issues/1136

torchtitan use cached state dict for ft. reset_sharded_param should be idempotent if model.parameters() are padded already

```
# pad DTensor._local_tensor
fully_shard(model)
sd = fsdp_model.state_dict()
# reset_sharded_param should be a no-op in lazy_init
loss = fsdp_model(inp).sum()
```

this PR make `reset_sharded_param` idempotent by checking storage data ptr and return early

unit test
```
pytest -s test/distributed/_composable/fsdp/test_fully_shard_state_dict.py -k test_cached_state_dict
```



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

weifengpy commented:
@pytorchmergebot merge

wwwjn commented:
> model.state_dict() -> model(input) reset_sharded_param should be no-op

Thanks for the fix @weifengpy ! By reading the code and comments, the `reset_sharded_param()` should only pad again if the model local tensor changed, right?
